tests/bluetooth/tester/src/btp_ots.c

```diff
@@ -0,0 +1,336 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/bluetooth/services/ots.h>
+
+#include <zephyr/sys/byteorder.h>
+#include <stdint.h>
+
+#include <zephyr/logging/log.h>
+#define LOG_MODULE_NAME bttester_ots
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_BTTESTER_LOG_LEVEL);
+
+#include "btp/btp.h"
+
+#define OBJ_POOL_SIZE CONFIG_BT_OTS_MAX_OBJ_CNT
+#define OBJ_MAX_SIZE  100
+
+static struct object {
+	uint8_t data[OBJ_MAX_SIZE];
+	char name[CONFIG_BT_OTS_OBJ_MAX_NAME_LEN + 1];
+	bool in_use;
+} objects[OBJ_POOL_SIZE];
+
+struct object_creation_data {
+	struct object *object;
+	struct bt_ots_obj_size size;
+	uint32_t props;
+};
+
+#define OTS_OBJ_ID_TO_OBJ_IDX(id) (((id) - BT_OTS_OBJ_ID_MIN) % ARRAY_SIZE(objects))
+
+static struct object_creation_data *object_being_created;
+
+static struct bt_ots *ots;
+
+static uint8_t ots_supported_commands(const void *cmd, uint16_t cmd_len,
+				      void *rsp, uint16_t *rsp_len)
+{
+	struct btp_ots_read_supported_commands_rp *rp = rsp;
+
+	tester_set_bit(rp->data, BTP_OTS_READ_SUPPORTED_COMMANDS);
+	tester_set_bit(rp->data, BTP_OTS_REGISTER_OBJECT);
+
+	*rsp_len = sizeof(*rp) + 1;
+
+	return BTP_STATUS_SUCCESS;
+}
+
+static struct object *get_object(void)
+{
+	for (unsigned int i = 0; i < ARRAY_SIZE(objects); i++) {
+		if (!objects[i].in_use) {
+			objects[i].in_use = true;
+			return &objects[i];
+		}
+	}
+
+	return NULL;
+}
+
+static uint8_t register_object(const void *cmd, uint16_t cmd_len,
+			       void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_ots_register_object_cmd *cp = cmd;
+	struct btp_ots_register_object_rp *rp = rsp;
+	struct object_creation_data obj_data;
+	struct bt_ots_obj_add_param param;
+	uint32_t supported_props = 0;
+	struct object *obj;
+	uint32_t props;
+	int err;
+
+	if ((cmd_len < sizeof(*cp)) || (cmd_len != sizeof(*cp) + cp->name_len)) {
+		return BTP_STATUS_FAILED;
+	}
+
+	if (cp->name_len == 0 || cp->name_len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
+		return BTP_STATUS_FAILED;
+	}
+
+	/* all supported props (execute, append, truncate not supported) */
+	BT_OTS_OBJ_SET_PROP_DELETE(supported_props);
+	BT_OTS_OBJ_SET_PROP_READ(supported_props);
+	BT_OTS_OBJ_SET_PROP_WRITE(supported_props);
+	BT_OTS_OBJ_SET_PROP_PATCH(supported_props);
+
+	props = sys_le32_to_cpu(cp->ots_props);
+	if (cp->flags & BTP_OTS_REGISTER_OBJECT_FLAGS_SKIP_UNSUPPORTED_PROPS) {
+		props &= supported_props;
+	}
+
+	obj = get_object();
+	if (!obj) {
+		return BTP_STATUS_FAILED;
+	}
+
+	(void)memset(&obj_data, 0, sizeof(obj_data));
+
+	memcpy(obj->name, cp->name, cp->name_len);
+	obj_data.object = obj;
+	obj_data.size.cur = sys_le32_to_cpu(cp->current_size);
+	obj_data.size.alloc = sys_le32_to_cpu(cp->alloc_size);
+	obj_data.props = props;
+
+	object_being_created = &obj_data;
+
+	param.size = obj_data.size.alloc;
+	param.type.uuid.type = BT_UUID_TYPE_16;
+	param.type.uuid_16.val = BT_UUID_OTS_TYPE_UNSPECIFIED_VAL;
+
+	err = bt_ots_obj_add(ots, &param);
+	object_being_created = NULL;
```

Comment:
I guess you are using `object_being_created` as a global variable since the API doesn't have a `user_data` or similar, right? 

It is probably worth a comment :)

Reply:
yeah,  I have a feeling that OTS could have few API tweaks here and there

Reply:
Indeed. Needing a global variable to act as user_data is not only poor design, but against our API guidelines :D